### ISSUE TITLE
Code needs a lot of refactoring

### ISSUE THREAD

**Issue by apatel762** [apatel762/static-site-generator#11]
I want to add unit tests (#10) but the code is pretty hard to test at the moment and could do with some refactoring to make things easier. Also the files and folders are organised strangely, so would be good to fix that up while I move stuff about during the refactoring.

My idea is that all of the note-processing logic should be split out into different 'Hook' classes which all implement a common set of functions (e.g. `BacklinkScannerHook` for scanning the notes for backlinks or `StateFileManagerHook` for managing the state file etc.). Then I'd do something like:

```python
for note in notes:
    for hook in hooks:
        hook.apply(note)
```

and as I add more features, I can just create new hooks and they'll be picked up somehow by the main loop.

Just an idea at this point, still needs some thinking about, but I think if I do this, it'll make things easier to test, as I can just write tests for each hook separately, instead of having one huge test class.

**Comment by apatel762** [comment]
[Helpful article](https://jeffknupp.com/blog/2013/08/16/open-sourcing-a-python-project-the-right-way/) re: organising a Python project properly

**Comment by apatel762** [comment]
Closing this as I've decided to rework the entire system.

I originally made this project as a way of injecting backlinks into my notes and then generating a static site from those notes, but going forward I'm going to separate those two things.

I'll be creating a separate project to handle injecting backlinks into the raw markdown notes and then I'll probably just use an existing static site generator to turn the notes into a website.
